### PR TITLE
release: promote develop to main (router security, coverage)

### PR DESCRIPTION
**Pull request by ZappoMan**
### Summary

This promotion brings three commits to production since CalVer tag **2026.005**: a security patch for React Router, targeted code-quality fixes in setup and mobile help flows, and restored 100% package test coverage after the Vitest 4 upgrade. The headline change for adopters is upgrading `react-router-dom` to **6.30.4**, which addresses open-redirect vulnerability **CVE-2026-40181**. There are no new database migrations, required secrets, or pending npm changesets in this release.

### Highlights

- **Security:** Upgrade `react-router-dom` to 6.30.4 across web, admin, and articles packages (CVE-2026-40181) (#398)
- **Mobile:** Scope lazy-loaded `ProfileEditor` state to the component instead of module-level mutable state (#399)
- **Setup & help:** Use path-based main-module detection in `setup-full.mjs`; parse help markdown synchronously via `marked.lexer` / `marked.parser` (#399)
- **Testing / CI:** Restore 100% package coverage under Vitest 4 with targeted tests for async cancellation, dynamic import fallbacks, and edge-case branches in billing, admin, connections, observability, kit, shared, and waitlist packages (#400)

**Stats:** 3 commits, +453 / −42 lines across 21 files.

### Adopter notes

| Area                       | Action |
| -------------------------- | ------ |
| **Secrets**                | None new |
| **Database**               | No new migrations |
| **Likely merge conflicts** | `package-lock.json`, `apps/web/package.json`, `packages/admin/package.json`, `packages/articles/package.json` (react-router bump); fork-customized `scripts/setup-full.mjs`, `apps/mobile/src/screens/ProfileScreen.tsx` |
| **npm packages**           | None — no pending changesets |

- **Landing:** https://beakerstack.com
- **Quick start:** https://github.com/Artificer-Innovations/BeakerStack/blob/main/QUICKSTART.md
- **Feedback:** https://github.com/Artificer-Innovations/BeakerStack/discussions
- **Upgrade guide:** https://github.com/Artificer-Innovations/BeakerStack/blob/main/docs/UPGRADING.md

---

### Merge instructions

**Use “Create a merge commit” — do NOT squash.** Required by [CONTRIBUTING.md](../../CONTRIBUTING.md) and enforced by [check-merge-strategy.yml](../workflows/check-merge-strategy.yml).

### Post-merge

1. Monitor **Deploy Production** workflow on `main`
2. Review auto-opened **“chore: release packages”** PR from changesets (if any); merge if correct
3. Run **Release Template** workflow on `main` (uses this PR’s CalVer section + git-cliff changelog)

### Test plan

- [ ] CI green on `develop` at promotion SHA
- [ ] Staging smoke: auth, dashboard, billing
- [ ] After merge: production deploy succeeds
- [ ] After **Release Template**: GitHub Release body matches **CalVer release notes** above + changelog

### CODE REVIEW

File: packages/admin/src/components/AdminDetailDrawer.web.test.tsx [hunk 1]
```diff
@@ -307,4 +307,19 @@ describe('AdminDetailDrawer', () => {
...
+        <button type='button'>Save</button>
+      </AdminDetailDrawer>
+    );
+    const notes = screen.getByLabelText('Notes');
+    notes.focus();
+    document.dispatchEvent(
+      new KeyboardEvent('keydown', { key: 'Tab', bubbles: true })
+    );
+    expect(document.activeElement).toBe(notes);
+  });
```
[the remark this answers is not in the document]
Comment (ZappoMan) [reply packages/admin/src/components/AdminDetailDrawer.web.test.tsx]: Addressed in #402 (merged to `develop`).

Replaced raw `KeyboardEvent` dispatch with `userEvent.tab()` and assert focus moves Notes → Save, verifying Tab is not trapped when focus is not on the last element. Hardened further in f7e2714 on that PR.

File: packages/waitlist/src/hooks/useSignupMode.test.tsx [hunk 2]
```diff
@@ -18,4 +18,27 @@ describe('useSignupMode', () => {
...
+    const supabase = {} as never;
+    const { unmount } = renderHook(() => useSignupMode(supabase));
+    await waitFor(() =>
+      expect(client.getPublicWaitlistSettings).toHaveBeenCalled()
+    );
+    unmount();
+    resolveSettings({ signup_mode: 'open', copy: {}, metadata_schema: [] });
+    await Promise.resolve();
+    vi.restoreAllMocks();
+  });
```
[the remark this answers is not in the document]
Comment (ZappoMan) [reply packages/waitlist/src/hooks/useSignupMode.test.tsx:59]: Addressed in #402 (merged to `develop`).

Added `console.error` spy + `expectNoUnmountedConsoleWarnings()` after resolving settings post-unmount. f7e2714 wrapped the spy in `try/finally` and scans all console.error args so the assertion is leak-proof and stricter than first-arg-only matching.

File: packages/billing/src/BillingProvider.test.tsx [hunk 3]
```diff
@@ -585,4 +585,105 @@ describe('BillingProvider', () => {
...
+    const { unmount } = render(
+      <BillingProvider config={testBillingConfig} {...providerProps}>
+        <PlanReader />
+      </BillingProvider>
+    );
+    await waitFor(() => expect(db.planMaybeSingle).toHaveBeenCalled());
+    unmount();
+    resolvePlan({ data: testPlan({ display_name: 'Late plan' }), error: null });
+    await Promise.resolve();
+  });
```
[the remark this answers is not in the document]
Comment (ZappoMan) [reply packages/billing/src/BillingProvider.test.tsx:637]: Addressed in #402 (merged to `develop`).

Plan query resolve-after-unmount test now spies on `console.error` and asserts no unmounted-component warnings via `expectNoUnmountedConsoleWarnings()`, with `act()` to flush updates. f7e2714 added `try/finally` cleanup and all-arg warning matching.

File: packages/billing/src/BillingProvider.test.tsx [hunk 4]
```diff
@@ -585,4 +585,105 @@ describe('BillingProvider', () => {
...
+      </BillingProvider>
+    );
+    await waitFor(() =>
+      expect(screen.getByTestId('uid').textContent).toBe('u-rpc-unmount')
+    );
+    await waitFor(() => expect(db.rpc).toHaveBeenCalled());
+    unmount();
+    resolveRpc({ error: null });
+    await Promise.resolve();
+  });
```
[the remark this answers is not in the document]
Comment (ZappoMan) [reply packages/billing/src/BillingProvider.test.tsx:669]: Addressed in #402 (merged to `develop`).

RPC resolve-after-unmount test now asserts no `console.error` unmounted warnings after late resolution. Same `expectNoUnmountedConsoleWarnings()` helper + `try/finally` spy cleanup from f7e2714.

File: packages/billing/src/BillingProvider.test.tsx [hunk 5]
```diff
@@ -585,4 +585,105 @@ describe('BillingProvider', () => {
...
+    const { unmount } = render(
+      <BillingProvider config={testBillingConfig} {...providerProps}>
+        <PlanReader />
+      </BillingProvider>
+    );
+    await waitFor(() => expect(db.planMaybeSingle).toHaveBeenCalled());
+    unmount();
+    rejectPlan(new Error('late plan fail'));
+    await Promise.resolve();
+  });
```
[the remark this answers is not in the document]
Comment (ZappoMan) [reply packages/billing/src/BillingProvider.test.tsx:704]: Addressed in #402 (merged to `develop`).

Plan query rejection-after-unmount test now spies on `console.error` and fails on unmounted-component warnings. f7e2714 ensures spy restore on assertion failure and matches warnings across all console.error arguments.

File: packages/billing/src/BillingProvider.test.tsx [hunk 6]
```diff
@@ -585,4 +585,105 @@ describe('BillingProvider', () => {
...
+      </BillingProvider>
+    );
+    await waitFor(() =>
+      expect(screen.getByTestId('uid').textContent).toBe('u-rpc-err-unmount')
+    );
+    await waitFor(() => expect(db.rpc).toHaveBeenCalled());
+    unmount();
+    rejectRpc(new Error('late rpc fail'));
+    await Promise.resolve();
+  });
```
[the remark this answers is not in the document]
Comment (ZappoMan) [reply packages/billing/src/BillingProvider.test.tsx:736]: Addressed in #402 (merged to `develop`).

RPC rejection-after-unmount test now has the same explicit postcondition: no unmounted warnings in `console.error`. Hardened in f7e2714 with `try/finally` and full-argument scanning.

**Review by diego-artificerinnovations — APPROVED** [review]
Reviewed all four commits. No blockers. Approving for promotion.

**`react-router-dom` 6.30.3 → 6.30.4 (CVE-2026-40181)** ✓ — Correct bump across all three consumers (apps/web, packages/admin, packages/articles). `package-lock.json` changes are consistent.

**`ProfileScreen.tsx` — Module-level state → component state** ✓ — Moving `ProfileEditor` from a module-level `let` to `useState` is the right fix. Module-level mutable state persists across remounts and hot-reloads; component state is correctly scoped. `setProfileEditor(() => module.ProfileEditor)` uses the functional-update form to avoid treating the component type as an initializer. Effect deps updated to `[isEditing, ProfileEditor]` — correct.

**`build-help.ts` — `marked.parser(marked.lexer(body))` over `marked.parse() as string`** ✓ — `marked.parse()` returns `string | Promise<string>` depending on configuration; the cast was a latent runtime hazard. `lexer` + `parser` are both synchronous with no overload ambiguity. Correct fix.

**`setup-full.mjs` — main-module detection** ✓ — `process.argv[1] && path.resolve(process.argv[1]) === __filename` is cleaner than the old `url.pathToFileURL(process.argv[1] || '').href` form, which had a latent bug: `pathToFileURL('')` returns the cwd URL, not the script URL, so an empty `argv[1]` could give a false positive. The new form explicitly guards the empty case. Assuming `__filename` is declared at the top of the file via `fileURLToPath(import.meta.url)` (standard ESM pattern; CI would catch it immediately if missing).

**`useUsage.ts` — ref init `undefined` + optional chain** ✓ — `useRef(undefined)` with `fetchUsageRef.current?.()` is more honest than a no-op stub; it correctly represents the pre-assignment state and avoids silently swallowing subscription events before `fetchUsage` is wired.

**Coverage tests (#400) and Copilot fixes (#402)** ✓ — Already reviewed and approved individually.

Adopter notes in the PR body look accurate (no new migrations, no new secrets, correct merge-conflict callouts).
